[quant][graph] Add check for qconfig_dict key

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37125 [quant][graph] Add JIT passes for dynamic quant multi uses of quant node
* #37093 [quant][graph] Run dynamic quantization for specific ops
* #37045 [graph][quant] Enable accessing child/grandchild modules in forward
* **#37014 [quant][graph] Add check for qconfig_dict key**

Summary:
User should only pass name as key in dict.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21283696](https://our.internmc.facebook.com/intern/diff/D21283696)

supriyar committed:
[quant][graph] API to preprocess qconfig_dict
Summary:
User should be able to pass module type or name as a key in qconfig dict.
if the qconfig_dict contains type as key, user should call `preprocess_qconfig_dict` function before calling prepare
If name is specified for a module then that takes precedence over type.

Works for both static and dynamic

Test Plan:
python test/test_quantize_script.py test_qconfig_w_types
python test/test_quantize_script.py test_dynamic_qconfig_w_types

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1159,6 +1161,20 @@ def forward(self, x):
                    .check("CallMethod") \
                    .run(model.graph)
 
+    def test_qconfig_w_types(self):
```
it's good enough to test the expected qconfig_dict I think

jerryzh168 replied:
can you add a comment here to say that this test exercises the config by name takes precedence over config by type?

jerryzh168 replied:
also could you change the test name to `test_qcong_with_types`?

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1794,6 +1810,39 @@ def forward(self, x):
         FileCheck().check("quantized::linear_dynamic") \
                    .run(model.graph)
 
+    def test_dynamic_qconfig_w_types(self):
```
same for this one, we can have end to end tests in some other places

jerryzh168 replied:
also there is probably nothing special for dynamic regarding this function, so you don't need to replicate this test for dynamic, unless you are doing integration test

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -80,13 +90,24 @@ def get_scripted_qconfig_dict(qconfig_dict):
 
 def _prepare_script(model, qconfig_dict, is_dynamic):
     _check_is_script_module(model)
+    module_types = list(filter(lambda x : isinstance(x, type), qconfig_dict.keys()))
```
you can use `any` I think

jerryzh168 replied:
also, probably check they are string, since we only accept string.

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -62,6 +62,16 @@ def __setstate__(self, state):
+        if type(child) in module_types:
+            dict_w_names[module_prefix] = qconfig_dict[type(child)]
+        else:
+            dict_w_names.update(_preprocess_qconfig_dict_helper(child, module_types, qconfig_dict, module_prefix))
```
nit: dict_w_names -> dict_with_names
also could you pass this as an argument and update the dictionary directly?

jerryzh168 replied:
actually could you just use qconfig_dict?

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -62,6 +62,16 @@ def __setstate__(self, state):
+    dict_w_names = {}
+    for name, child in module.named_children():
+        module_prefix = prefix + '.' + name if prefix else name
+        if type(child) in module_types:
```
why do you need to get module types? can't just look up this in qconfig_dict?

supriyar replied:
Yeah, will fix this

jerryzh168 reviewed (changes requested):
had some comments

supriyar committed:
Update on "[quant][graph] API to preprocess qconfig_dict"


Summary:
User should be able to pass module type or name as a key in qconfig dict.
if the qconfig_dict contains type as key, user should call `preprocess_qconfig_dict` function before calling prepare
If name is specified for a module then that takes precedence over type.

Works for both static and dynamic

Test Plan:
python test/test_quantize_script.py test_qconfig_w_types
python test/test_quantize_script.py test_dynamic_qconfig_w_types

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] API to preprocess qconfig_dict"


Summary:
User should be able to pass module type or name as a key in qconfig dict.
if the qconfig_dict contains type as key, user should call `preprocess_qconfig_dict` function before calling prepare
If name is specified for a module then that takes precedence over type.

Works for both static and dynamic

Test Plan:
python test/test_quantize_script.py test_qconfig_w_types
python test/test_quantize_script.py test_dynamic_qconfig_w_types

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -64,6 +64,16 @@ def __setstate__(self, state):
     conv_packed_params = torch.jit.script(ConvPackedParams())._c
 
+def _preprocess_qconfig_dict_helper(module, qconfig_dict, prefix=''):
+    dict_with_names = {}
```
can you just use qconfig_dict here? and filter out the types in the end?

supriyar replied:
I'm using qconfig_dict to access the original qconfigs in the parent function (since that take precedence over anything set here). I just thought this is cleaner to read.

jerryzh168 replied:
I feel it's pretty clear to just use qconfig_dict, but remember copy the qconfig_dict before you modify it.

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -82,13 +92,24 @@ def get_scripted_qconfig_dict(qconfig_dict):
 
 def _prepare_script(model, qconfig_dict, is_dynamic):
     _check_is_script_module(model)
+    if not all(map(lambda x : isinstance(x, str), qconfig_dict.keys())):
```
nit: this can be any + not isinstance

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -64,6 +64,16 @@ def __setstate__(self, state):
+    dict_with_names = {}
+    for name, child in module.named_children():
+        module_prefix = prefix + '.' + name if prefix else name
+        if type(child) in qconfig_dict.keys():
```
@supriyar I think you can do the precedence check here

supriyar committed:
Update on "[quant][graph] API to preprocess qconfig_dict"


Summary:
User should be able to pass module type or name as a key in qconfig dict.
if the qconfig_dict contains type as key, user should call `preprocess_qconfig_dict` function before calling prepare
If name is specified for a module then that takes precedence over type.

Works for both static and dynamic

Test Plan:
python test/test_quantize_script.py test_qconfig_w_types
python test/test_quantize_script.py test_dynamic_qconfig_w_types

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_quantize_script.py:
```diff
@@ -82,13 +92,23 @@ def get_scripted_qconfig_dict(qconfig_dict):
                                                                False,
                                                                is_dynamic))
 
+def preprocess_qconfig_dict(module, qconfig_dict):
```
Does this need to be a top level API? Can this be called inside prepare_script?

supriyar replied:
We need to pass in the module here to be able to access child module. In prepare script we take scripted module so we lose child information

jerryzh168 replied:
the child info is still there, we lose the type info in scripted module

jerryzh168 replied:
same for this function, qconfig_dict should be the first argument.

raghuramank100 commented on torch/quantization/_quantize_script.py:
```diff
@@ -82,13 +92,23 @@ def get_scripted_qconfig_dict(qconfig_dict):
 
 def _prepare_script(model, qconfig_dict, is_dynamic):
     _check_is_script_module(model)
+    if not all(map(lambda x : isinstance(x, str), qconfig_dict.keys())):
```
Can we run preprocess_qconfig_dict automatically as part of _prepare_script for dynamic quant?

supriyar replied:
answered the question above. We can change the prepare api to accept module instance as well. But that is not very clean since prepare inputs and outputs are scripted module.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1158,6 +1160,16 @@ def forward(self, x):
 
+    def test_qconfig_with_types(self):
+        data = [(torch.rand((1, 5), dtype=torch.float), torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        qconfig_dict = {torch.nn.Linear : default_qconfig, 'sub2.fc1' : default_per_channel_qconfig}
```
Is this precedence also enforced for the eager API?

supriyar replied:
I don't see the logic to enforce this in eager API

raghuramank100 replied:
Can you do that as a follow up PR so that eager and graph are consistent?

jerryzh168 replied:
yeah precedence is enforced here in eager: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L41-L43

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -64,6 +64,16 @@ def __setstate__(self, state):
+        # Existing string names in dict take precedence over type names.
+        if type(child) in qconfig_dict.keys() and module_prefix not in qconfig_dict.keys():
+            qconfig_dict[module_prefix] = qconfig_dict[type(child)]
+        else:
+            qconfig_dict.update(_preprocess_qconfig_dict_helper(child, qconfig_dict, module_prefix))
```
this can be removed now?

jerryzh168 replied:
oh I mean, we don't need to update now

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -64,6 +64,16 @@ def __setstate__(self, state):
+            qconfig_dict[module_prefix] = qconfig_dict[type(child)]
+        else:
+            qconfig_dict.update(_preprocess_qconfig_dict_helper(child, qconfig_dict, module_prefix))
+    return qconfig_dict
```
also I think we should make this function a inplace function

jerryzh168 replied:
also please remember to change _preprocess_qconfig_dict_helper to `_preprocess_qconfig_dict_helper_` and qconfig should be the first argument of this function because that is the primary object we are preprocessing.

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -82,13 +92,23 @@ def get_scripted_qconfig_dict(qconfig_dict):
+def preprocess_qconfig_dict(module, qconfig_dict):
+    dict_has_types = any(map(lambda x : isinstance(x, type), qconfig_dict.keys()))
+    if dict_has_types:
+        qconfig_dict = _preprocess_qconfig_dict_helper(module, qconfig_dict)
```
we can copy the qconfig_dict here and make a inplace modification in the function.

supriyar committed:
Update on "[quant][graph] API to preprocess qconfig_dict"


Summary:
User should be able to pass module type or name as a key in qconfig dict.
if the qconfig_dict contains type as key, user should call `preprocess_qconfig_dict` function before calling prepare
If name is specified for a module then that takes precedence over type.

Works for both static and dynamic

Test Plan:
python test/test_quantize_script.py test_qconfig_w_types
python test/test_quantize_script.py test_dynamic_qconfig_w_types

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -82,6 +82,8 @@ def get_scripted_qconfig_dict(qconfig_dict):
 def _prepare_script(model, qconfig_dict, is_dynamic):
     _check_is_script_module(model)
+    if any(map(lambda x : not isinstance(x, str), qconfig_dict.keys())):
+        raise ValueError('qconfig_dict should contain names(str) as keys.')
```
I think this should be an illegalargument exception?

raghuramank100 replied:
A general check (like we have in eager mode) would be to warn if a qconfig_dict causes no change to the model, this means that the keys in the dict do not map to modules in the model, we have an error message in eager code at: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L142

supriyar committed:
Update on "[quant][graph] Add check for qconfig_dict key"


Summary:
User should only pass name as key in dict.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
